vignettes/python-primer.Rmd

```diff
@@ -0,0 +1,920 @@
+prefix to a symbol. (You'll see the `*` prefix again when we talk about
+functions)
+
+```{python}
+x = (1, 2, 3)
+a, *the_rest = x
+a
+the_rest
+```
+
+#### Dictionaries
+
+Dictionaries are most similar to R's named lists. They are a container
+where you can retrieve items by name, though in Python the name (called
+a *key* in Python's parlance) does not need to be a string like in R. It
+can be any Python object with a `hash()` method (meaning, it can be
+almost any Python object). They can be created using syntax like
+`{key: value}`. Like Python lists, they are modified in place.
+
+```{python}
+d = {"key1": 1,
+     "key2": 2}
+d2 = d
+d
+d["key1"]
+d["key3"] = 3
+d2 # modified in place!
+```
+
+Unlike R's named lists, you cannot index into a dictionary with an
+integer to get an item at a specific index position. Dictionaries are
+*unordered* containers. (However---beginning with Python 3.7,
+dictionaries do preserve the item insertion order).
+
+```{python, error = TRUE}
+d = {"key1": 1, "key2": 2}
+d[1] # error
+```
+
+A container that closest matches the semantics of R's named list is the
+`OrderedDict`, but that's relatively uncommon in Python code so we don't
+cover it further.
+
+#### Sets
+
+Sets are a container that can be used to efficiently track unique items
+or deduplicate lists. They are constructed using `{val1, val2}` (unlike
+a dictionary, without `:`). Think of them as dictionary where you only
+use the keys. Sets have many efficient methods for membership
+operations, like `intersection()`, `issubset()`, `union()` and so on.
+
+```{python}
+s = {1, 2, 3}
+type(s)
+s
+
+s.add(1)
+s
+```
+
+### Iteration with `for`
+
+The `for` statement in Python can be used to iterate over any kind of
+container.
+
+```{python}
+for x in [1, 2, 3]:
+  print(x)
+```
+
+R has a relatively limited set of objects that can be passed to `for`.
+Python by comparison, provides an iterator protocol interface, which
+means that authors can define custom objects, with custom behavior that
+is invoked by `for`. (We'll have an example for how to define a custom
+iterable when we get to classes). You may want to use a Python iterable
+from R using reticulate, so it's helpful to peel back the syntactic
+sugar a little to show what the `for` statement is doing in Python, and
+how you can step through it manually.
+
+There are two things that happen: first, an iterator is constructed from
+the supplied object. Then, the new iterator object is repeatedly called
+with `next()` until it is exhausted.
+
+```{python, error = TRUE}
+l = [1, 2, 3]
+it = iter(l) # create an iterator object
+it
+
+# call `next` on the iterator until it is exhausted:
+next(it)
+next(it)
+next(it)
+next(it)
+```
+
+In R, you can use reticulate to step through an iterator the same way.
+
+```{r}
+library(reticulate)
+l <- r_to_py(list(1, 2, 3))
+it <- as_iterator(l)
+
+iter_next(it)
+iter_next(it)
+iter_next(it)
+iter_next(it, completed = "StopIteration")
```

Comment:
Maybe also mention the `coro::loop` syntax here?

Reply:
Thanks! I added a mention about coro::loop in a later "final remarks on iteration" section.